src/core/main/common/FreezeResolver.ts

```diff
@@ -0,0 +1,533 @@
+      this._decipherabilityFreezeStartingTimestamp = null;
+      return null;
+    }
+
+    const freezingTs = freezing?.timestamp ?? rebuffering?.timestamp ?? null;
+    log.info("FR: Freeze detected", freezingTs, now - (freezingTs ?? now));
+
+    /**
+     * If `true`, we recently tried to "flush" to unstuck playback but playback
+     * is still stuck
+     */
+    const recentFlushAttemptFailed =
+      this._lastFlushAttempt !== null &&
+      now - this._lastFlushAttempt.timestamp < FREEZING_FLUSH_FAILURE_DELAY.MAXIMUM &&
+      now - this._lastFlushAttempt.timestamp >= FREEZING_FLUSH_FAILURE_DELAY.MINIMUM &&
+      Math.abs(freezingPosition - this._lastFlushAttempt.position) <
+        FREEZING_FLUSH_FAILURE_DELAY.POSITION_DELTA;
+
+    if (recentFlushAttemptFailed) {
+      const secondUnfreezeStrat = this._getStrategyIfFlushingFails(freezingPosition);
+      this._decipherabilityFreezeStartingTimestamp = null;
+      this._ignoreFreezeUntil = now + 6000;
+      return secondUnfreezeStrat;
+    }
+
+    const decipherabilityFreezeStrat = this._checkForDecipherabilityRelatedFreeze(
+      observation,
+      now,
+    );
+    if (decipherabilityFreezeStrat !== null) {
+      return decipherabilityFreezeStrat;
+    }
+
+    if (freezingTs !== null && now - freezingTs > UNFREEZING_SEEK_DELAY) {
+      this._lastFlushAttempt = {
+        timestamp: now,
+        position: freezingPosition + UNFREEZING_DELTA_POSITION,
+      };
+      log.debug("FR: trying to flush to un-freeze");
+
+      this._decipherabilityFreezeStartingTimestamp = null;
+      this._ignoreFreezeUntil = now + 6000;
+      return {
+        type: "flush",
+        value: { relativeSeek: UNFREEZING_DELTA_POSITION },
+      };
+    }
+    return null;
+  }
+
+  /**
+   * Performs decipherability-related checks if it makes sense.
+   *
+   * If decipherability-related checks have been performed **AND** an
+   * un-freezing strategy has been selected by this method, then return
+   * an object describing this wanted unfreezing strategy.
+   *
+   * If this method decides to take no action for now, it returns `null`.
+   * @param {Object} observation - playback observation that has just been
+   * performed.
+   * @param {number} now - Monotonically-raising timestamp for the current
+   * time.
+   * @returns {Object|null}
+   */
+  private _checkForDecipherabilityRelatedFreeze(
+    observation: IFreezeResolverObservation,
+    now: number,
+  ): IFreezeResolution | null {
+    const { readyState, rebuffering, freezing, fullyLoaded } = observation;
+    const bufferGap = normalizeBufferGap(observation.bufferGap);
+    const rebufferingForTooLong =
+      rebuffering !== null && now - rebuffering.timestamp > 4000;
+    const frozenForTooLong = freezing !== null && now - freezing.timestamp > 4000;
+
+    const hasDecipherabilityFreezePotential =
+      (rebufferingForTooLong || frozenForTooLong) &&
+      ((bufferGap < 6 && !fullyLoaded) || readyState > 1);
```

Comment:
You're right this parenthesis looks reversed it should maybe be `(bufferGap >= 6 || fullyLoaded)`....

Will look at the previous code because it must be a refactoring mistake.

Reply:
It was a refactoring mistake: it was code I moved from the old code to the new but I should have reversed the conditions.
Done